libs/components/packages/src/schematics/migrations/update-7/ag-grid-28.schematic.ts

```diff
@@ -0,0 +1,195 @@
+import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
+import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
+import {
+  NodeDependencyType,
+  addPackageJsonDependency,
+  getPackageJsonDependency,
+  removePackageJsonDependency,
+} from '@schematics/angular/utility/dependencies';
+
+import ignore from 'ignore';
+import { relative } from 'path';
+import semver from 'semver';
+
+export const UPDATE_TO_VERSION = '28.2.0';
+
+const UPDATE_LT_VERSION = '28.0.0';
+const ANY_MODULE = '@ag-grid-community/';
+const ENT_MODULE = '@ag-grid-enterprise/';
+const AG_GRID = 'ag-grid-community';
+const AG_GRID_ENT = 'ag-grid-enterprise';
+const withComponents = 'AgGridModule.withComponents';
+const warned: string[] = [];
+
+function warnOnce(message: string) {
+  if (!warned.includes(message)) {
+    warned.push(message);
+    console.warn(message);
+  }
+}
+
+/**
+ * Upgrade to AG Grid 28 and address three breaking changes:
+ *
+ * - No longer support `@ag-grid-community/all-modules` -- just use the package
+ * - Remove `AgGridModule.withComponents` in module import
+ * - Column API renamed getSecondaryColumns / setSecondaryColumns
+ *
+ * Also advise against mixing modules and packages.
+ */
+export default function (): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    let needsPackageUpdated = false;
+    let needsEntPackageUpdated = false;
+    const agGridAllModules = getPackageJsonDependency(
+      tree,
+      `${ANY_MODULE}all-modules`
+    );
+    const agGridDependency = getPackageJsonDependency(tree, AG_GRID);
+    const agGridAllModulesEnt = getPackageJsonDependency(
+      tree,
+      `${ENT_MODULE}all-modules`
+    );
+    const agGridDependencyEnt = getPackageJsonDependency(tree, AG_GRID_ENT);
+    let type = NodeDependencyType.Default;
+    const overwrite = true;
+    const version = UPDATE_TO_VERSION;
+    if (!agGridAllModules && !agGridDependency) {
+      // AG Grid is not installed, so we don't need to do anything.
+      return;
+    }
+
+    // Determine if there are any community packages that need to be updated.
+    if (
+      agGridDependency &&
+      semver.lt(agGridDependency.version, UPDATE_LT_VERSION)
+    ) {
+      type = agGridDependency.type;
+      needsPackageUpdated = true;
+    } else if (!agGridDependency && agGridAllModules) {
+      type = agGridAllModules.type;
+      needsPackageUpdated = true;
+    }
+    if (needsPackageUpdated) {
+      [AG_GRID, 'ag-grid-angular'].forEach((name) => {
+        addPackageJsonDependency(tree, { name, overwrite, type, version });
+      });
+    }
+    if (agGridAllModules) {
+      removePackageJsonDependency(tree, `${ANY_MODULE}all-modules`);
+      removePackageJsonDependency(tree, `${ANY_MODULE}angular`);
+    }
+
+    // Determine if there are any enterprise packages that need to be updated.
+    if (
+      agGridDependencyEnt &&
+      semver.lt(agGridDependencyEnt.version, UPDATE_LT_VERSION)
```

Comment:
Instead of adding a new dependency (`semver`) and checking if the version is less than the desired version, could we _always_ update the version listed in `package.json` (even if it's the same)? Even though it might add some minimal time to the update, it'll reduce the complexity of the operation considerably. 

Something similar to: https://github.com/blackbaud/skyux/blob/main/libs/components/packages/src/schematics/rules/install-angular-cdk.ts